grai-integrations/source-bigquery/src/grai_source_bigquery/loader.py

```diff
@@ -40,14 +48,207 @@ def get_from_env(
+        self.dataset = get_from_env("dataset", required=False) if dataset is None else dataset
+        self.credentials = get_from_env("credentials", required=False) if credentials is None else credentials
+        self.window = int(get_from_env("window", required=False, default=7)) if window is None else window
+        self._connection: Optional[logging.Client] = None
+
+    def __enter__(self):
+        return self.connect()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def connect(self) -> "LoggingConnector":
+        """
+
+        Args:
+
+        Returns:
+
+        Raises:
+
+        """
+        if self._connection is None:
+            credentials = None
+
+            if self.credentials:
+                json_acct_info = json.loads(self.credentials, strict=False)
+                credentials = service_account.Credentials.from_service_account_info(json_acct_info)
+
+            self._connection = logging.Client(credentials=credentials, project=self.project)
+        return self
+
+    @property
+    def connection(self) -> logging.Client:
+        """
+
+        Args:
+
+        Returns:
+
+        Raises:
+
+        """
+        if self._connection is None:
+            raise Exception("Not connected, call `.connect()")
+        return self._connection
+
+    def close(self) -> None:
+        """
+
+        Args:
+
+        Returns:
+
+        Raises:
+
+        """
+        self.connection.close()
+        self._connection = None
+
+    @cached_property
+    def logs(self) -> List[Any]:
+        """Create and return a list of dictionaries with the
+        schemas and names of tables in the database
+        connected to by the connection argument.
+
+        Args:
+
+        Returns:
+
+        Raises:
+
+        """
+
+        yesterday = datetime.now(timezone.utc) - timedelta(days=self.window)
+        time_format = "%Y-%m-%dT%H:%M:%S.%f%z"
+
+        filter_str = (
+            'protoPayload.serviceName="bigquery.googleapis.com"'
+            ' AND resource.type = "bigquery_project"'
+            ' AND protoPayload.methodName="google.cloud.bigquery.v2.JobService.InsertJob"'
+            f' AND timestamp>="{yesterday.strftime(time_format)}"'
+        )
+
+        return self.connection.list_entries(filter_=filter_str, page_size=1000)
+
+    def get_edges(self, existing_nodes) -> List[Edge]:
+        """
+
+        Args:
+
+        Returns:
+
+        Raises:
+
+        """
+
+        def table_string_to_table_id(table_string: str) -> Optional[TableID]:
+            table_string = table_string.split("/")
+
+            schema = table_string[3]
+            name = table_string[5]
+
+            if not any(node.name == name and node.table_schema == schema for node in existing_nodes):
+                return None
+
+            return TableID(
+                table_schema=schema,
+                name=name,
+                namespace=self.namespace,
+            )
+
+        edges = set()
```

Comment:
This is a good idea in principle but TableID objects won't hash this way. The hash will end up being on the `id` of the instance effectively making every instance of a Table unique. You can still do this but you'll have to add an `__hash__` method to `TableID` with something like


```python

def __hash__(self):
    return hash((self.name, self.namespace))

```

Reply:
I have indeed put a `__hash__` method on `Edge`

Reply:
Point 1 is covered by line 166 `if not any(node.name == name and node.table_schema == schema for node in existing_nodes):`

The set is trying to cover point 2, at least from the point of view of this integration. I agree there is a chance there will already be an edge between these sources and destinations, but isn't that often the case? For example a dbt edge and a BigQuery edge? in which case should it be the responsibility of the merge logic to sort this out?

Reply:
> Point 1 is covered by line 166 if not any(node.name == name and node.table_schema == schema for node in existing_nodes):

Ahh, thank you. I see it now.

> The set is trying to cover point 2, at least from the point of view of this integration. I agree there is a chance there will already be an edge between these sources and destinations, but isn't that often the case?

I might have done a bad job of describing the scenario because I'm talking about something different.

| Dataset | schema | name | full_name |
| -------- | ------- | ------ | --------- |
| a | test | name | **test.name** |
| b | test | name | **test.name** |

So far as the edge is concerned nothing currently disambiguates between the node in a and b. Depending on execution order all of the following are possible: 

1. An edge intended to be between `test.name` in dataset (a) and another table in (a) results in an edge between  `test.name` in (b) and the intended other table in (a).
2. Either valid edges corresponding to (a) and (b) will be omitted (one gobbles the other in the set) or the DB errors when duplicate full_names are provided in the update. The latter scenario would happen if the source/destination specs were correctly identified as belonging to different nodes.

Reply:
I think you can fix this with the Fivetran namespace trick - allow the user to do the disambiguation by mapping dataset -> namespace. 

I suppose the real question is whether we ought to error when we've detected an invalid arrangement (like two nodes we know are different but have the same (name, namespace)), risk silently inserting bad data, or something else.

Reply:
Small enhancement to line 166 as well so it's not an O(n) operation.

```python
        existing_node_unique_ids = {(node.name, node.table_schema) for node in existing_nodes}
        
        def table_string_to_table_id(table_string: str) -> Optional[TableID]:
            table_string = table_string.split("/")

            schema = table_string[3]
            name = table_string[5]

            if (name, schema) not in existing_node_unique_ids:
                return None
```

Reply:
`dataset` and `schema` are the same thing in BigQuery, hence I don't think this is an issue.